spec/formatting.md

```diff
@@ -18,3 +18,71 @@ the local variable takes precedence.
 
 It is an error for a local variable definition to
 refer to a local variable that's defined after it in the message.
+
+## Error Handling
+
+During the formatting of a message,
+various errors may be encountered.
+These are divided to the following categories:
+
+- **Syntax errors** occur when the syntax representation of a message is invalid.
+- **Resolution errors** occur when the runtime value of a part of a message
+  cannot be determined.
+
+  - **Unresolved Variable** errors occur when a variable reference cannot be resolved.
+
+- **Selection errors** cover failures encountered during selection.
+
+  - **Selector errors** are failures in the matching of a key to a specific selector.
+  - **Missing Fallback** errors occur when no Variant is selected
+    due to the message not including a Variant with only catch-all keys.
+
+- **Formatting errors** occur during the formatting of a resolved value,
+  for example when encountering a value with an unsupported type
+  or an internally inconsistent set of options.
+
+During selection, an expression handler must only emit Resolution and Selection errors.
+During formatting, an expression handler must only emit Resolution and Formatting errors.
+
+In all cases, when encountering an error,
+a message formatter must provide some representation of the message.
+An informative error or errors must also be separately provided.
+
+When an error occurs in the syntax or resolution of an Expression or MarkupStart Option,
+the Expression or MarkupStart in question is processed as if the option was not defined.
+This may allow for the fallback handling described below to be avoided,
+though an error must still be emitted.
+
+When an error occurs within a Selector,
+the selector must not match any VariantKey other than the catch-all `*`
+and a Selector error is emitted.
+When selection fails to match any Variant,
+an empty string is used as the formatted string representation of the message
+and a Missing Fallback error is emitted.
```

Comment:
See our previous discussion. In terms of a separation of concerns, the Selector should decide what it does internally. What MF probably needs to specify (as I wrote in a comment above) is that the Selectors *must* always select one of the available VariantKeys and *should* select `*` when selection-internal errors are encountered.

Note that line 60 mixes selection and formatting "modes". If these are separate (which I think they need to be) that line should say instead:

> an empty string is used as the result of message selection

Actually, we might have multiple selection going on, so we can't really say that. The Selector *must* choose an available VariantKey. If none can be matched (such as Missing Fallback error), then the Selector can emit what amounts to a fatal error *for its own operation* and the formatter can be directed to emit the empty string. Does that make sense?

Note that "missing fallback" can be detected by the MF parser before selection even happens--we _could_ make it a syntax error (actually, I thought we did?). Your text on lines 59-61 says it is not a syntax error and only throws Missing Fallback if no values match.

Reply:
Missing an ultimate fallback case is indeed a [syntax error](https://github.com/unicode-org/message-format-wg/blob/main/spec/syntax.md#variants):

> A well-formed message is considered valid if the following requirements are satisfied:
> - The number of keys on each variant must be equal to the number of selectors.
> - At least one variant's keys must all be equal to the catch-all key (`*`).

However, I think we still need this here to account for what happens when an MF2 data model representation of a message has been formed by some other means which might be more lenient.

Reply:
> "other means which might be more lenient"

But why?
Parse time is probably the best time to check this.
Parsing is something that can be done statically, so the more we catch there, the better it is.
Other types of errors can only be done at runtime (unresolved variables, functions, bad type)

Even if some implementation is lenient, we should still consider this a syntax error, not some other kind of error.

Reply:
We cannot always consider this a syntax error at runtime, because the syntax for the message might not be MF2.

Let's say that there's a FooFormat9000 that's exactly like MF1, except that it doesn't require an "other" case for its selectors and defaults to an empty string instead. Now, if a FooFormat9000 message is parsed into an MF2 data model for use with an MF2 runtime, the code doing so might not add the expected default variant to the data model. When the MF2 runtime discovers this, it needs to emit a Selection Error rather than a Syntax Error, because it doesn't know anything about FooFormat9000 where this syntax is valid.

If the message was originally in MF2 syntax, then we would of course have also emitted a Syntax Error earlier.

Reply:
Resolving this thread, as the PR has been updated to include Data Model errors as a separate category.